content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md

```diff
@@ -0,0 +1,545 @@
+supports the required compute capability. Then load an appropriate CCE version.
+For information about which CCE versions support which versions of CUDA, see
+the [HPE Cray Programming Environment release announcements](https://cpe.ext.hpe.com/docs/latest/release_announcements/index.html). If in doubt, the default versions on your system
+should work just fine.
+
+When writing this tutorial, I used CCE 17 and CUDA 12 and ran the application on NVIDIA A100 GPUs, but the debugging process is the same for any compiler combination.
+
+After deciding which versions you want, use the `module` command in a terminal to ensure that the Cray `CC` compiler and the NVIDIA `nvcc` compiler are available.
+
+```shell
+$ module load cce/17
+$ module load cuda/12
+```
+
+You can check that your environment is set up correctly by checking that CC
+(the Cray CCE C++ compiler) and nvcc (the NVIDIA CUDA application compiler)
+are available in the terminal.
+
+```shell
+$ CC --version
+Cray clang version 17.0.1  (5ec9405551a8c8845cf14e81dc28bff7aa3935cb)
+...
+
+$ nvcc --version
+nvcc: NVIDIA (R) Cuda compiler driver
+Copyright (c) 2005-2024 NVIDIA Corporation
+Built on Thu_Jun__6_02:18:23_PDT_2024
+Cuda compilation tools, release 12.5, V12.5.82
+...
+```
+
+#### Compile the Application
+
+The sample application comes with a Makefile, but it's not compatible with the
+HPE Cray Programming Environment. You will compile the application manually.
+
+#### Debug Info
+
+When compiling an application, compilers apply optimizations to make the resulting
+application faster and smaller. These optimizations often rearrange the order
+of code or completely remove it which makes reasoning about the original code
+at runtime difficult. When preparing to debug an application, you need to tell
+the compiler to disable these optimizations and include extra information that
+the debugger can use to reason about the original code.
+
+#### Compile the .cu File
+
+The simpleMPI.cu file contains the code that will be run on the GPU. Use
+nvcc to compile it.
+
+```shell
+$ nvcc -g -G -O0 -c -ccbin=cc -gencode arch=... simpleMPI.cu -o simpleMPI_gpu.o
+```
+
+The `-g` flag tells nvcc to include debugging information for the host code
+(the code that will run on the CPU) and the `-G` flag tells nvcc to include
+debugging information for the device code (the code that will run on the GPU).
+`-O0` disables optimizations for the host code. `-G` implicitly disables
+optimizations for the device code.
+
+`-c` tells the compiler to produce an object file instead of a executable. An
+object file is a mostly-compiled program that only needs to be linked to create
+an executable. After you compile the .cpp file in the next section, you will
+link the object files together to create the final executable.
+
+nvcc does not actually fully compile C/C++. It actually processes the CUDA
```

Comment:
This would be another great place to say "The nvcc [something] does not actually" so as to not start the sentence and paragraph with a lower case letter. (I'm not exactly sure what to call it, though. I'll leave that up to you.)